infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/workflows/HealNodeWorkflow.java

```diff
@@ -89,4 +92,39 @@ public void impl(@Nonnull CorfuRuntime runtime) throws Exception {
             newLayout = new Layout(runtime.getLayoutView().getLayout());
         }
     }
+
+    /**
+     * This action attempts to restore redundancy for all servers across all segments
+     * starting from the oldest segment. It then collapses the segments once the set of
+     * servers in the 2 oldest subsequent segments are equal.
+     */
+    class RestoreRedundancyAndMergeSegments extends Action {
+        @Nonnull
+        @Override
+        public String getName() {
+            return "RestoreRedundancyAndMergeSegments";
+        }
+
+        @Override
+        public void impl(@Nonnull CorfuRuntime runtime) throws Exception {
+            while (newLayout.getSegments().size() > 1) {
+
+                // Catchup all servers.
+                Set<String> lowRedundancyServers = Sets.difference(
```

Comment:
This is not correct . You are catching up all servers from all stripes.But actually it should be only all servers per stripe ? Or are we saying replicas are per segment ?

Reply:
We have state transfer at the segment granularity and not for stripes.
Added comment.